cli/admin/call.go

```diff
@@ -0,0 +1,187 @@
+		"AdminCallFuncName": adminCallFuncName,
+		"FuncName":          funcName,
+		"Args":              argsSerialized,
+	})
+
+	callResRaw, err := common.EvalTarantoolConn(conn, callFuncBody)
+	if err != nil {
+		return fmt.Errorf("Failed to call %q: %s", funcName, err)
+	}
+
+	needReturnValueWarn := false
+
+	switch callRes := callResRaw.(type) {
+	case string:
+		log.Info(callRes)
+	case []interface{}:
+		for _, callResLineRaw := range callRes {
+			switch callResLine := callResLineRaw.(type) {
+			case string:
+				log.Info(callResLine)
+			default:
+				needReturnValueWarn = true
+				log.Infof("%v\n", callResLine)
+			}
+		}
+	default:
+		needReturnValueWarn = true
+		log.Infof("%v\n", callRes)
+	}
+
+	if needReturnValueWarn {
+		log.Warnf("Admin function %q should return string or string array value", funcName)
+	}
+
+	return nil
+}
+
+func getFuncCallArgs(conn net.Conn, funcName string, flagSet *pflag.FlagSet, args []string) ([]FuncCallArg, error) {
+	helpResRawMap, err := getFuncHelpRawMap(funcName, conn)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get function %q signature: %s", funcName, err)
+	}
+
+	argsSpec, err := getArgsSpec(helpResRawMap)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get %q arguments spec: %s", funcName, err)
+	}
+
+	conflictingFlagNames := getConflictingFlagNames(argsSpec, flagSet)
+	if len(conflictingFlagNames) > 0 {
+		return nil, fmt.Errorf(
+			"Function has arguments with names that conflicts with `cartridge admin` flags: %s",
+			strings.Join(conflictingFlagNames, ", "),
+		)
+	}
+
+	funcCallArgs := make([]FuncCallArg, len(argsSpec))
+
+	i := 0
+	for argName, argSpec := range argsSpec {
+		funcCallArgs[i] = FuncCallArg{
+			Name: argName,
+			Type: argSpec.Type,
+		}
+
+		switch argSpec.Type {
+		case argTypeString:
+			flagSet.StringVar(&funcCallArgs[i].StringValue, argName, "", argSpec.Usage)
+		case argTypeNumber:
+			flagSet.IntVar(&funcCallArgs[i].NumberValue, argName, 0, argSpec.Usage)
+		case argTypeBool:
+			flagSet.BoolVar(&funcCallArgs[i].BoolValue, argName, false, argSpec.Usage)
+		default:
+			return nil, fmt.Errorf(
+				"Admin function %q accepts value of unsupported type: %s",
+				funcName, argSpec.Type,
+			)
+		}
+
+		i++
+	}
+
+	flagSet.SetNormalizeFunc(func(f *pflag.FlagSet, name string) pflag.NormalizedName {
+		return pflag.NormalizedName(normalizeFlagName(name))
+	})
+
+	flagSet.ParseErrorsWhitelist = pflag.ParseErrorsWhitelist{
+		UnknownFlags: false,
+	}
+	if err := flagSet.Parse(args); err != nil {
+		return nil, fmt.Errorf("Failed to parse %q function arguments: %s", funcName, err)
+	}
+
+	// set Changed for args
+	for i := range funcCallArgs {
+		flag := flagSet.Lookup(funcCallArgs[i].Name)
+		if flag == nil {
+			return nil, project.InternalError("Flag %q isn't found", funcCallArgs[i].Name)
+		}
+
+		funcCallArgs[i].Changed = flag.Changed
+	}
+
+	return funcCallArgs, nil
+}
+
+func serializeArgs(funcCallArgs []FuncCallArg) (string, error) {
```

Comment:
I think it could me moved in some small library inside cli.

Reply:
Not sure - it's very specific serialization that works only with `[]FuncCallArg ` that is used only for admin call.